tsinfer/formats.py

```diff
@@ -2004,33 +2058,92 @@ def num_alleles(self, sites=None):
+        :rtype: iter(Variant)
+
+        .. warning::
+            If ``recode_ancestral=True``, the order of the alleles may not be the same
+            as in the ``sites.alleles`` list, and the indexes in the genotypes array
+            may change accordingly.
+
+        """
+        if recode_ancestral is None:
+            recode_ancestral = False
         all_genotypes = chunk_iterator(self.sites_genotypes, indexes=sites)
-        for genotypes, site in zip(all_genotypes, self.sites(ids=sites)):
-            yield Variant(site=site, alleles=site.alleles, genotypes=genotypes)
-
-    def __all_haplotypes(self, sites=None):
+        assert MISSING_DATA < 0  # required for geno_map to remap MISSING_DATA
+        for genos, site in zip(all_genotypes, self.sites(ids=sites)):
+            aa = site.ancestral_allele
+            alleles = site.alleles
+            if aa != MISSING_DATA and aa > 0 and recode_ancestral:
+                # Need to recode this site
+                alleles = (alleles[aa],) + alleles[:aa] + alleles[aa + 1 :]
+                # re-map the genotypes
+                geno_map = np.arange(len(alleles) - MISSING_DATA, dtype=genos.dtype)
+                geno_map[MISSING_DATA] = MISSING_DATA
+                geno_map[aa] = 0
+                geno_map[0:aa] += 1
+                genos = geno_map[genos]
+            yield Variant(site=site, alleles=alleles, genotypes=genos)
+
+    def __all_haplotypes(self, sites=None, recode_ancestral=None):
         # We iterate over chunks vertically here, and it's not worth complicating
         # the chunk iterator to handle this.
+        if recode_ancestral is None:
+            recode_ancestral = False
+        aa_index = self.sites_ancestral_allele[:]
+        # If ancestral allele is missing, keep the order unchanged (aa_index of zero)
+        aa_index[aa_index == MISSING_DATA] = 0
         chunk_size = self.sites_genotypes.chunks[1]
         for j in range(self.num_samples):
             if j % chunk_size == 0:
                 chunk = self.sites_genotypes[:, j : j + chunk_size].T
             a = chunk[j % chunk_size]
+            if recode_ancestral:
+                # Remap the genotypes at all sites, depending on the aa_index
+                a = np.where(
+                    a == aa_index,
+                    0,
+                    np.where(np.logical_and(a != MISSING_DATA, a < aa_index), a + 1, a),
+                )
             yield j, a if sites is None else a[sites]
 
-    def haplotypes(self, samples=None, sites=None):
+    def haplotypes(self, samples=None, sites=None, recode_ancestral=None):
         """
-        Returns an iterator over the (sample_id, haplotype) pairs.
+        Returns an iterator over the (sample_id, haplotype) pairs. Each haplotype is
+        an array of indexes, where the ``i``th value is an index into the
+        ``site.alleles`` list for the ``i`` specified site. A value of ``0`` in the
```

Comment:
Same comments as docstring for variants above. Could just do a "See variants for details" to avoid duplication.